### ISSUE TITLE
CSS: Not enough space before bullet lists

### ISSUE THREAD

**Issue by mgeier** [jupyter/notebook#1392]
Probably the existing rule from `renderedhtml.less` should be slightly altered:

```css
.rendered_html ul:not(.list-inline), .rendered_html ol:not(.list-inline) {
    margin: 1em 2em 0em 2em;
    padding-left: 0px;
}
```

This is an example where the problem is obvious:

```markdown
### Title
* one
* two
```

BTW, this is how it looks on Github:

### Title
* one
* two

**Comment by ellisonbg** [comment]
Our choice might be deliberate - but it could be a mistake. In general, our
visual style tends to be more compressed vertically.

On Tue, Apr 26, 2016 at 4:05 AM, Matthias Geier <notifications@github.com>
wrote:

> Probably the existing rule from renderedhtml.less should be slightly
> altered:
>
> .rendered_html ul:not(.list-inline), .rendered_html ol:not(.list-inline) {
>     margin: 1em 2em 0em 2em;
>     padding-left: 0px;
> }
>
> This is an example where the problem is obvious:
>
> ### Title* one* two
>
> BTW, this is how it looks on Github:
> Title
>
>    - one
>    - two
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/jupyter/notebook/issues/1392>
>



-- 
Brian E. Granger
Associate Professor of Physics and Data Science
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com

**Comment by fperez** [comment]
Stumbled upon this by accident, coincidentally this had been bugging me a few days ago. It seems like it got vertically tighter at some point, b/c I don't recall feeling it was *too* tight in the past, and recently it did seem to me like it was. Here's a side-by-side github/us comparison:

![image](https://cloud.githubusercontent.com/assets/57394/14838187/9c6b18a0-0bd3-11e6-8b80-bcb8a29f6ce0.png)

While I know we want a tight vertical layout, we might want to add a bit of space here, as I think our current choice is perhaps *too* tight. It's more noticeable with a longer title and more items in the list:

![image](https://cloud.githubusercontent.com/assets/57394/14838212/dda32466-0bd3-11e6-952f-bcde92d65fd7.png)

**Comment by ellisonbg** [comment]
Hmm, yes, I completely agree. Maybe we can see how/when/why it changed.
IIRC this spacing used to be very carefully tuned to the leading of the
text to match the vertical spacing of the type in regular paragraphs.

On Tue, Apr 26, 2016 at 5:25 PM, Fernando Perez <notifications@github.com>
wrote:

> Stumbled upon this by accident, coincidentally this had been bugging me a
> few days ago. It seems like it got vertically tighter at some point, b/c I
> don't recall feeling it was *too* tight in the past, and recently it did
> seem to me like it was. Here's a side-by-side github/us comparison:
>
> [image: image]
> <https://cloud.githubusercontent.com/assets/57394/14838187/9c6b18a0-0bd3-11e6-8b80-bcb8a29f6ce0.png>
>
> While I know we want a tight vertical layout, we might want to add a bit
> of space here, as I think our current choice is perhaps *too* tight. It's
> more noticeable with a longer title and more items in the list:
>
> [image: image]
> <https://cloud.githubusercontent.com/assets/57394/14838212/dda32466-0bd3-11e6-952f-bcde92d65fd7.png>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/jupyter/notebook/issues/1392#issuecomment-214928262>
>



-- 
Brian E. Granger
Associate Professor of Physics and Data Science
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com

**Comment by mgeier** [comment]
This is either a mistake in The Notebook, or a bug in Less or in Firefox and Chromium. It doesn't look deliberate to me. The change happened more than 2 years ago in 4363aa72f815545c5c57bc472a38f6cad3824d15, which was created by you, @ellisonbg.

You changed `margin: 1em 2em;` to `margin: 0em 2em;`. But you didn't forget the `margin-top`!
You also added 

```css
* + ul { margin-top: 1em; }
* + ol { margin-top: 1em; }
```

The intention is clear, but for some reason this doesn't work for me (and apparently for @fperez either).

Any ideas how this can be fixed?

BTW, this has nothing to do with having it more or less tight.
If you want it tighter, just change the `1em` to something smaller for all relevant CSS rules (which would be a lot of work!).

**Comment by mgeier** [comment]
ping?

**Comment by fperez** [comment]
@ellisonbg, do you think one of the CalPoly student teams might have some cycles to take a crack at this?

**Comment by eskirk** [comment]
We had a fix for this, but it was veto'd due to the code being too expensive. If i recall we might need to add a class tag to some of these elements. 

Regardless I'll take a look at this right now, seems ezpz.

**Comment by ellisonbg** [comment]
OK, we do need to fix it though

On Thu, Jun 23, 2016 at 12:43 PM, Elliot Kirk <notifications@github.com>
wrote:

> We had a fix for this, but it was veto'd due to the code being too
> expensive. If i recall we might need to add a class tag to some of these
> elements.
>
> Regardless I'll take a look at this right now, seems ezpz.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/jupyter/notebook/issues/1392#issuecomment-228207995>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AABr0Locn2mLdevM3cXqKiP0ucZiBT7aks5qOwwngaJpZM4IP1FC>
> .
>



-- 
Brian E. Granger
Associate Professor of Physics and Data Science
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com

**Comment by eskirk** [comment]
@ellisonbg I was wrong, that was for another issue ¯\\\_(ツ)_/¯

**Comment by eskirk** [comment]
This should be fixed with #1564, just waiting on a merge

**Comment by mgeier** [comment]
That's kinda making it worse!

Or rather it somewhat improves the example above, but breaks other examples.

E.g., if a Markdown cell starts with a list, this introduces spurious spacing.

Also (but the fix suggested above has the same problem), now there is too much spacing in nested lists. Is that what the `list-inline` class is for?

**Comment by eskirk** [comment]
@mgeier I could use a `h<n> + ul` tag to only have spacing after headings. Would that be better?

**Comment by mgeier** [comment]
Inserting the additional space only after headings feels like a hack, and it would still produce too little space after equations, tables, images, ...

In the meantime I found out why the rules mentioned above (https://github.com/jupyter/notebook/issues/1392#issuecomment-215022995) don't work:

```css
* + ul { margin-top: 1em; }
* + ol { margin-top: 1em; }
```

They were simply overruled by:

```css
ol:not(.list-inline) {margin: 0em 2em; padding-left: 0px;}
```

I removed this bug in #1568, which I think is a slight improvement, but it's still broken.
Especially nested lists still look awkward. You can try something like this:

```markdown
* first item
  * first subitem
  * second subitem
  
  paragraph after second subitem
* second item

  paragraph inside second item
* third item
```

The real problem, however, isn't only list spacing. The spacing of `<p>` and many other elements just doesn't work well together.
Sadly, I don't know enough CSS-fu to produce a solution for the whole problem, I hope somebody else does that!
